### ISSUE TITLE
[Request] Option to select which monitor to toggle HDR for

### ISSUE THREAD

**Issue by JediRhymeTrix** [Codectory/HDR-Profile#4]
I always leave HDR enabled in Windows for my TV and would like to toggle HDR only for my primary monitor.  The tool toggles both displays together. Would it be possible to add an option to select the monitor(s) which need to be toggled?

Also, "Deactivate HDR"  in the tray icon menu doesn't seem to work. It doesn't even get highlighted when I hover the cursor over it and it cannot be clicked.

**Comment by Codectory** [comment]
Does the toggle in the app works while the toggle in the tray icon does not?

I will check, if it's possible to selected specific monitors. Does Windows offer this out of the box?

**Comment by JediRhymeTrix** [comment]
The toggle in the app works fine. Only the deactivate button in the tray icon doesn't work.

Yes, the HDR toggle in Windows is on a per-monitor basis.

**Comment by Codectory** [comment]
See newest release: https://github.com/Codectory/HDR-Profile/releases/tag/1.5.5

**Comment by JediRhymeTrix** [comment]
This version fails to launch. it appears in the task manager for a couple of seconds and then disappears.

Windows 10 19043.844

Tried with no anti-virus and Windows Defender real-time protection disabled.

Log contents:

```
VEGAS	HDRPProfile 1.5.5.0

#######################
07-03-2021 09:26:06 PM:Initializing...
07-03-2021 09:26:06 PM:Loading settings...
```

**Comment by Codectory** [comment]
Can you delete the HDRPRofile_Settings.xml file?

**Comment by JediRhymeTrix** [comment]
Tried that too. It just creates the file again everytime I try to run it.

Here is a crash log from event viewer:

```
Application: HDRProfile.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.NullReferenceException
   at HDRProfile.MonitorManager.GetActiveMonitors()
   at HDRProfile.MonitorManager..ctor(HDRProfile.HDRProfileSettings)
   at HDRProfile.HDRProfileHandler.InitializeMonitorManager()
   at HDRProfile.HDRProfileHandler.Initialize()

Exception Info: System.Reflection.TargetInvocationException
   at System.RuntimeTypeHandle.CreateInstance(System.RuntimeType, Boolean, Boolean, Boolean ByRef, System.RuntimeMethodHandleInternal ByRef, Boolean ByRef)
   at System.RuntimeType.CreateInstanceSlow(Boolean, Boolean, Boolean, System.Threading.StackCrawlMark ByRef)
   at System.Activator.CreateInstance(System.Type, Boolean)
   at System.RuntimeType.CreateInstanceImpl(System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo, System.Object[], System.Threading.StackCrawlMark ByRef)
   at System.Activator.CreateInstance(System.Type, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo, System.Object[])
   at System.Activator.CreateInstance(System.Type, System.Object[])
   at System.Xaml.Schema.SafeReflectionInvoker.CreateInstanceCritical(System.Type, System.Object[])
   at System.Xaml.Schema.XamlTypeInvoker.CreateInstance(System.Object[])
   at MS.Internal.Xaml.Runtime.ClrObjectRuntime.CreateInstance(System.Xaml.XamlType, System.Object[])
   at System.Xaml.XamlObjectWriter.Logic_CreateAndAssignToParentStart(MS.Internal.Xaml.Context.ObjectWriterContext)
   at System.Xaml.XamlObjectWriter.WriteEndObject()
   at System.Windows.Markup.WpfXamlLoader.TransformNodes(System.Xaml.XamlReader, System.Xaml.XamlObjectWriter, Boolean, Boolean, Boolean, System.Xaml.IXamlLineInfo, System.Xaml.IXamlLineInfoConsumer, MS.Internal.Xaml.Context.XamlContextStack`1<System.Windows.Markup.WpfXamlFrame>, System.Windows.Markup.IStyleConnector)
   at System.Windows.Markup.WpfXamlLoader.Load(System.Xaml.XamlReader, System.Xaml.IXamlObjectWriterFactory, Boolean, System.Object, System.Xaml.XamlObjectWriterSettings, System.Uri)
   at System.Windows.Markup.WpfXamlLoader.LoadBaml(System.Xaml.XamlReader, Boolean, System.Object, System.Xaml.Permissions.XamlAccessLevel, System.Uri)
   at System.Windows.Markup.XamlReader.LoadBaml(System.IO.Stream, System.Windows.Markup.ParserContext, System.Object, Boolean)

Exception Info: System.Reflection.TargetInvocationException
   at System.RuntimeTypeHandle.CreateInstance(System.RuntimeType, Boolean, Boolean, Boolean ByRef, System.RuntimeMethodHandleInternal ByRef, Boolean ByRef)
   at System.RuntimeType.CreateInstanceSlow(Boolean, Boolean, Boolean, System.Threading.StackCrawlMark ByRef)
   at System.Activator.CreateInstance(System.Type, Boolean)
   at System.RuntimeType.CreateInstanceImpl(System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo, System.Object[], System.Threading.StackCrawlMark ByRef)
   at System.Activator.CreateInstance(System.Type, System.Reflection.BindingFlags, System.Reflection.Binder, System.Object[], System.Globalization.CultureInfo, System.Object[])
   at System.Activator.CreateInstance(System.Type, System.Object[])
   at System.Xaml.Schema.SafeReflectionInvoker.CreateInstanceCritical(System.Type, System.Object[])
   at System.Xaml.Schema.XamlTypeInvoker.CreateInstance(System.Object[])
   at MS.Internal.Xaml.Runtime.ClrObjectRuntime.CreateInstance(System.Xaml.XamlType, System.Object[])
   at System.Xaml.XamlObjectWriter.Logic_CreateAndAssignToParentStart(MS.Internal.Xaml.Context.ObjectWriterContext)
   at System.Xaml.XamlObjectWriter.WriteStartMember(System.Xaml.XamlMember)
   at System.Windows.Markup.WpfXamlLoader.TransformNodes(System.Xaml.XamlReader, System.Xaml.XamlObjectWriter, Boolean, Boolean, Boolean, System.Xaml.IXamlLineInfo, System.Xaml.IXamlLineInfoConsumer, MS.Internal.Xaml.Context.XamlContextStack`1<System.Windows.Markup.WpfXamlFrame>, System.Windows.Markup.IStyleConnector)
   at System.Windows.Markup.WpfXamlLoader.Load(System.Xaml.XamlReader, System.Xaml.IXamlObjectWriterFactory, Boolean, System.Object, System.Xaml.XamlObjectWriterSettings, System.Uri)
   at System.Windows.Markup.WpfXamlLoader.LoadBaml(System.Xaml.XamlReader, Boolean, System.Object, System.Xaml.Permissions.XamlAccessLevel, System.Uri)
   at System.Windows.Markup.XamlReader.LoadBaml(System.IO.Stream, System.Windows.Markup.ParserContext, System.Object, Boolean)
   at System.Windows.Application.LoadBamlStreamWithSyncInfo(System.IO.Stream, System.Windows.Markup.ParserContext)
   at System.Windows.Application.DoStartup()
   at System.Windows.Application.<.ctor>b__1_0(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.DispatcherOperation.InvokeImpl()
   at MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(System.Object)
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Windows.Threading.DispatcherOperation.Invoke()
   at System.Windows.Threading.Dispatcher.ProcessQueue()
   at System.Windows.Threading.Dispatcher.WndProcHook(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndWrapper.WndProc(IntPtr, Int32, IntPtr, IntPtr, Boolean ByRef)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(System.Object)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate, System.Object, Int32)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(System.Object, System.Delegate, System.Object, Int32, System.Delegate)
   at System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority, System.TimeSpan, System.Delegate, System.Object, Int32)
   at MS.Win32.HwndSubclass.SubclassWndProc(IntPtr, Int32, IntPtr, IntPtr)
   at MS.Win32.UnsafeNativeMethods.DispatchMessage(System.Windows.Interop.MSG ByRef)
   at System.Windows.Threading.Dispatcher.PushFrameImpl(System.Windows.Threading.DispatcherFrame)
   at System.Windows.Application.RunDispatcher(System.Object)
   at System.Windows.Application.RunInternal(System.Windows.Window)
   at HDRProfile.App.Main()
```

**Comment by Codectory** [comment]
Could you check if WMI service is running?

**Comment by JediRhymeTrix** [comment]
Yes, it is running.

**Comment by Codectory** [comment]
Can you try to run the app with administrative privileges?

Vedant Vohra <notifications@github.com> schrieb am So. 7. März 2021 um
17:43:

**Comment by JediRhymeTrix** [comment]
Tried it. No difference.

**Comment by atomizasser** [comment]
I also have the same problem. Previous version worked. Thanks

**Comment by Codectory** [comment]
Please try the following:

Press „ Windows + R“ -> Type „wbemtedt“ and press enter -> Connect -> Connect -> Query -> type „ Select * From Win32_DiplayMonitor“ -> press apply -> double click the first entry -> send me the pnpdevicetypeid

**Comment by JediRhymeTrix** [comment]
I'm getting an error with description: "Invalid Class". I've even tried querying "Win32_DisplayMonitor" instead of "Win32_DiplayMonitor".

**Comment by Codectory** [comment]
My bad, it's Win32_DesktopMonitor. I'm writing from my mobile phone as I am not at home. My mind isn'5 the best ^^

**Comment by JediRhymeTrix** [comment]
Yeah, was just about to update my comment that I tried that. Got a couple of results. Double clicked the first one but couldn't find pnpdevicetypeid but found PNPDeviceID instead, but it's <null>

The 2nd item had PNPDeviceID - DISPLAY\ACR06C1\5&24E74614&0&UID4355

Not sure if this is what you're looking for

**Comment by Codectory** [comment]
So the first one doesn‘t has the PNPDeviceID property? Can you send me a
screenshot of all the properties? This should be the issue. I assumed that
all devices have this id, which u use to identify each monitor.

Vedant Vohra <notifications@github.com> schrieb am So. 7. März 2021 um
19:49:

**Comment by JediRhymeTrix** [comment]
First one: ![First one](https://user-images.githubusercontent.com/26216205/110251030-8ead0580-7fa4-11eb-864a-d2e74ef643b2.png)

Second one: ![Second one](https://user-images.githubusercontent.com/26216205/110251052-aedcc480-7fa4-11eb-9dab-8a7d6ddff826.png)

**Comment by JediRhymeTrix** [comment]
They both have DeviceID maybe that can be used instead?

**Comment by Codectory** [comment]
How many entries do you have in the list and how many monitors do you have? I have to set the hdr mode in the C++ part and I am not that familiar with C++. I will check if there is another way to get the UID (part of pnpdeviceid) or other ways to identify the monitors.

**Comment by JediRhymeTrix** [comment]
Okay, this is the strange part... I have 3 monitors connected but the query only returns 2. Monitor1 in the list seems to be Monitor #3 and Monitor2 seems to be the primary (Monitor #1) as per display settings and Nvidia Control Panel. Monitor #2 seems to be missing from this list.

**Comment by Codectory** [comment]
Found a new method. Please, give it a a try: https://github.com/Codectory/HDR-Profile/releases/tag/1.5.6

**Comment by JediRhymeTrix** [comment]
1.5.6 is launching fine and the auto HDR behavior at a per monitor level is working perfectly so far. However, the HDR toggle in the tray icon menu isn't working yet. I guess the application thinks HDR is on when it is on for any one monitor, which could be causing the issue. Also, it would be awesome if the manual HDR toggle could also be done on a per monitor basis or at least the option to select which monitors to toggle when the manual activate/deactivate button is clicked.

**Comment by Codectory** [comment]
Glad to hear, the we actually fixed this!
The toggle should work as the button on the "Status" page in the app. When you click on the button, all monitors which have Auto HDR activated, will be toggled. Is this not working for you?

**Comment by JediRhymeTrix** [comment]
![image](https://user-images.githubusercontent.com/26216205/110434480-e009e180-80d7-11eb-9e49-ad324fdf068c.png)

As you can see, the only monitor with HDR on currently is one that has Auto HDR disabled. However, the application is showing "Deactivate HDR" on the button and when I click on the button, it does nothing. The same behavior is reflected in the tray icon toggle too.

![image](https://user-images.githubusercontent.com/26216205/110434670-1e070580-80d8-11eb-882c-6ebb112a4576.png)

**Comment by Codectory** [comment]
You did manually activate HDR for Display3 right? I need to ingore monitors, where Auto HDR is not activated. HDRProfile will not toggle them but reading the status and if HDR is enabled for one monitor, you cannot enable it. The problem is, you cannot disable it too, because Auto HDR is not activate for this monitor. I will fix this. For now, you can just activate Auto HDR for Display3

**Comment by JediRhymeTrix** [comment]
Just noticed that this has been fixed (for me, at least) in newer builds. Thanks a ton! 

@Codectory since you haven't updated this issue, I presume you're still testing? If so, please feel free to close this issue whenever you'd like. From my end, it seems to be resolved as of version 1.5.7.

**Comment by Codectory** [comment]
Should been fixed, correct. As you confirm it, I will close this issue.
